components/server/src/workspace/workspace-starter.ts

```diff
@@ -276,20 +278,24 @@ export class WorkspaceStarter {
             const ideConfig = await this.resolveIDEConfiguration(ctx, workspace, user, options.ideSettings);
 
             // create and store instance
-            let instance = await this.workspaceDb
-                .trace({ span })
-                .storeInstance(
-                    await this.newInstance(
-                        ctx,
-                        workspace,
-                        lastValidWorkspaceInstance,
-                        user,
-                        project,
-                        options.excludeFeatureFlags || [],
-                        ideConfig,
-                        options.workspaceClass,
-                    ),
-                );
+            let instance = await this.newInstance(
+                ctx,
+                workspace,
+                lastValidWorkspaceInstance,
+                user,
+                project,
+                options.excludeFeatureFlags || [],
+                ideConfig,
+                options.workspaceClass,
+            );
+            // we run the actual creation of a new instance in a distributed lock, to make sure we always only start one instance per workspace.
+            await this.synchronizer.synchronized("startws-" + workspace.id, "server", async () => {
```

Comment:
Yes, it would require writing a dedicated function in the DB layer (as I cannot pass down connections to workspacedb), but I believe it's a simpler approach in this case. Thanks for the suggestion, will give it a go.

Reply:
I looked deeper and think using the distributed lock is better:
 - mysql's default isolation level (REPEATABLE READ) doesn't help, as it would allow concurrent inserts of new data
 - the best alternative would be to use SELECT ... FOR UPDATE, which requires to explicitly maintain a larger SQL statement and keep it in sync with the workspaceinstance shape. Doable but more complicated.
 - the distributed lock is a primitive we have already and doesn't do any locking for other things that starting the particular workspace. It just calls two existing db operations.

Reply:
Okay, just really feels like a big hammer for something that should be solveable with a transaction, given we are only trying to maintain the invariant that there are at most 1 WSI for each WS, and we know the WS ID